lib/Subscription/Captcha.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+namespace MailPoet\Subscription;
+
+use MailPoet\Models\Subscriber;
+use MailPoet\Models\SubscriberIP;
+use MailPoet\Util\Helpers;
+use MailPoet\WP\Functions as WPFunctions;
+use MailPoetVendor\Gregwar\Captcha\CaptchaBuilder;
+
+class Captcha {
+  const TYPE_BUILTIN = 'built-in';
+  const TYPE_RECAPTCHA = 'recaptcha';
+  const TYPE_DISABLED = null;
+
+  const SESSION_KEY = 'mailpoet_captcha';
+  const SESSION_FORM_KEY = 'mailpoet_captcha_form';
+
+  /** @var WPFunctions */
+  private $wp;
+
+  function __construct(WPFunctions $wp = null) {
+    if ($wp === null) {
+      $wp = new WPFunctions;
+    }
+    $this->wp = $wp;
+  }
+
+  function isSupported() {
+    return extension_loaded('gd') && function_exists('imagettftext');
+  }
+
+  function isRequired($subscriber_email = null) {
+    if ($this->wp->isUserLoggedIn()) {
+      return false;
+    }
+
+    // Check limits per recipient
+    $subscription_captcha_recipient_limit = $this->wp->applyFilters('mailpoet_subscription_captcha_recipient_limit', 1);
+    if ($subscriber_email) {
+      $subscriber = Subscriber::where('email', $subscriber_email)->findOne();
+      if ($subscriber instanceof Subscriber
+        && $subscriber->count_confirmations >= $subscription_captcha_recipient_limit
+      ) {
+        return true;
+      }
+    }
+
+    // Check limits per IP address
+    $subscription_captcha_window = $this->wp->applyFilters('mailpoet_subscription_captcha_window', MONTH_IN_SECONDS);
+
+    $subscriber_ip = Helpers::getIP();
+
+    if (!empty($subscriber_ip)) {
```

Comment:
We can get rid of one level nesting conditions by immediately returning `false` `if (empty($subscriber_ip))`